### PR TITLE
[BEAM-8298] Implement side input caching.

### PR DESCRIPTION
**Pull request by robertwb**
State tokens are supported, as well as a fallback to caching within bundles.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by robertwb** [comment]
CC: @mxm

File: sdks/python/apache_beam/runners/worker/sdk_worker.py [hunk 1]
```diff
@@ -759,50 +759,57 @@ def __init__(self,
...
       if cache_token_struct.HasField("user_state"):
         # There should only be one user state token present
         assert not user_state_cache_token
         user_state_cache_token = cache_token_struct.token
+      elif cache_token_struct.HasField("side_input"):
+        self._context.side_input_cache_tokens[
+            cache_token_struct.side_input.transform_id,
+            cache_token_struct.side_input.side_input_id
+        ] = cache_token_struct.token
+    self._context.bundle_cache_token = bundle_id
```
Comment (mxm) [inline sdks/python/apache_beam/runners/worker/sdk_worker.py]: Bundle caching for get requests is a good improvement.

File: sdks/python/apache_beam/runners/worker/sdk_worker.py [hunk 2]
```diff
@@ -825,43 +833,92 @@ def extend(self,
...
+    data, continuation_token = self._underlying.get_raw(state_key, None)
+    head = []
+    input_stream = coder_impl.create_InputStream(data)
+    while input_stream.size() > 0:
+      head.append(coder.decode_from_stream(input_stream, True))
+
+    if continuation_token is None:
+      return head
+    else:
+      def iter_func():
```
Comment (mxm) [inline sdks/python/apache_beam/runners/worker/sdk_worker.py]: AFAIK Continuation tokens are not implemented but it's good we support this now.

File: sdks/python/apache_beam/runners/worker/statecache.py [hunk 3]
```diff
@@ -148,47 +148,43 @@ class StateCache(object):
...
-      token, value = self._cache.get(state_key)
-    return value if token == cache_token else None
+      return self._cache.get((state_key, cache_token))
 
   @Metrics.counter("put")
   def put(self, state_key, cache_token, value):
     assert cache_token and self.is_cache_enabled()
     with self._lock:
-      return self._cache.put(state_key, (cache_token, value))
+      return self._cache.put((state_key, cache_token), value)
```
Comment (mxm) [inline sdks/python/apache_beam/runners/worker/statecache.py]: The combined use of `(state_key, cache_token)` simplifies the lookup. Thanks for fixing this.

**Comment by lukecwik** [comment]
StateCacheTest failures need to be resolved.

**Comment by robertwb** [comment]
Fixed statecache tests and added a unit test.

**Comment by lukecwik** [comment]
Seeing test failures:
```
13:11:10 AttributeError: '_IterableFromIterator' object has no attribute 'extend'
```

and linter problems:
```
12:54:20 > Task :sdks:python:test-suites:tox:py37:lintPy37
12:54:20 ************* Module apache_beam.runners.worker.sdk_worker_test
12:54:20 apache_beam/runners/worker/sdk_worker_test.py:162:0: C0330: Wrong hanging indentation (add 2 spaces).
12:54:20       token=b'state_token1',
12:54:20       ^ | (bad-continuation)
12:54:20 apache_beam/runners/worker/sdk_worker_test.py:163:0: C0330: Wrong hanging indentation (add 2 spaces).
12:54:20       user_state=beam_fn_api_pb2.ProcessBundleRequest.CacheToken.UserState())
12:54:20       ^ | (bad-continuation)
12:54:20 apache_beam/runners/worker/sdk_worker_test.py:165:0: C0330: Wrong hanging indentation (add 2 spaces).
12:54:20       token=b'state_token2',
12:54:20       ^ | (bad-continuation)
12:54:20 apache_beam/runners/worker/sdk_worker_test.py:166:0: C0330: Wrong hanging indentation (add 2 spaces).
12:54:20       user_state=beam_fn_api_pb2.ProcessBundleRequest.CacheToken.UserState())
12:54:20       ^ | (bad-continuation)
12:54:20 apache_beam/runners/worker/sdk_worker_test.py:168:0: C0330: Wrong hanging indentation (add 2 spaces).
12:54:20       token=b'side1_token1',
12:54:20       ^ | (bad-continuation)
12:54:20 apache_beam/runners/worker/sdk_worker_test.py:169:0: C0330: Wrong hanging indentation (add 2 spaces).
12:54:20       side_input=beam_fn_api_pb2.ProcessBundleRequest.CacheToken.SideInput(
12:54:20       ^ | (bad-continuation)
12:54:20 apache_beam/runners/worker/sdk_worker_test.py:172:0: C0330: Wrong hanging indentation (add 2 spaces).
12:54:20       token=b'side1_token2',
12:54:20       ^ | (bad-continuation)
12:54:20 apache_beam/runners/worker/sdk_worker_test.py:173:0: C0330: Wrong hanging indentation (add 2 spaces).
12:54:20       side_input=beam_fn_api_pb2.ProcessBundleRequest.CacheToken.SideInput(
12:54:20       ^ | (bad-continuation)
```

File: sdks/python/apache_beam/runners/worker/sdk_worker_test.py [hunk 4]
```diff
@@ -115,6 +118,116 @@ def test_fn_registration(self):
     self._check_fn_registration_multi_request((1, 4), (4, 4))
 
 
+class CachingStateHandlerTest(unittest.TestCase):
```
Comment (lukecwik) [inline sdks/python/apache_beam/runners/worker/sdk_worker_test.py]: Nice.

**Comment by robertwb** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by robertwb** [comment]
Pushed changes fixing the flink tests.

**Pull request merged**